pkg/glusterutils/cache.go

```diff
@@ -0,0 +1,281 @@
+package glusterutils
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+// GCache is a wrapper around 'GInterface' object
+type GCache struct {
+	gd                 GInterface
+	ttl                time.Duration
+	lock               sync.Mutex
+	lastCallValueMap   map[string]interface{}
+	lastCallTimeMap    map[string]time.Time
+	cacheDisabledFuncs map[string]struct{}
+}
+
+// NewGCacheWithTTL method creates a new GCache wrapper instance
+func NewGCacheWithTTL(gd GInterface, ttl time.Duration) *GCache {
+	var gc = new(GCache)
+	gc.gd = gd
+	gc.ttl = 1 * time.Minute // default to 1 minute
+	gc.SetTTL(ttl)
+	gc.lastCallValueMap = make(map[string]interface{})
+	gc.lastCallTimeMap = make(map[string]time.Time)
+	gc.cacheDisabledFuncs = make(map[string]struct{})
+	return gc
+}
+
+// NewGCache method creates a new GCache wrapper instance,
+// with 1 minute default time_to_live
+func NewGCache(gd GInterface) *GCache {
+	return NewGCacheWithTTL(gd, 1*time.Minute)
+}
+
+// TTL method returns the current time_to_live duration
+func (gc *GCache) TTL() time.Duration {
+	return gc.ttl
+}
+
+// SetTTL method sets a new time_to_live
+func (gc *GCache) SetTTL(ttl time.Duration) {
+	// accepts only if the given ttl is greater than a second
+	if ttl > time.Second {
+		gc.ttl = ttl
+	}
+}
+
+// DisableCacheForFuncs method will disable the caching functionalities
+// for the provided list of functions.
+// If the provided function is not there in the existing list, it will be ignored
+func (gc *GCache) DisableCacheForFuncs(fNames []string) {
+	for _, fName := range fNames {
+		gc.cacheDisabledFuncs[fName] = struct{}{}
+	}
+}
+
+func (gc *GCache) timeForNewCall(funcName string, origFuncName string) (ret bool) {
+	if origFuncName == "" {
+		origFuncName = funcName
+	}
+	ret = true
+	// if the function is disabled, always return true
+	if _, ok := gc.cacheDisabledFuncs[origFuncName]; ok {
+		return
+	}
+	if _, ok := gc.lastCallTimeMap[funcName]; !ok {
+		return
+	}
+	nowT := time.Now()
+	// if the last called time is BEFORE 'now - ttl', then call again
+	if gc.lastCallTimeMap[funcName].Before(nowT.Add(-gc.ttl)) {
+		return
+	}
+	return false
+}
+
+// EnableVolumeProfiling method wraps the GInterface.EnableVolumeProfiling call
+func (gc *GCache) EnableVolumeProfiling(vInfo Volume) error {
```

Comment:
Yes @shtripat , current implementation (lets call it cache-0.1v) has the limitation.

But this can be easily scaled by adding ```go```'s reflection mechanisms, all the redundant codes can be clubbed together and even new functions can be added without touching ```cache.go``` file.

We can add these functionalities before the next release.

Reply:
I would say if we go this PR way, lets try to get the reflection thing also in place. @aravindavk comments ???

Reply:
Since this PR is also not avoiding an extra function for each exposed interface function. Can we proceed with the PR #132 and enhance the caching with this PR later?

@shtripat @aruniiird

Reply:
@aravindavk I would agree to this proposal and take this PR later with reflection thing in place.

Reply:
> Since this PR is also not avoiding an extra function for each exposed interface function. Can we proceed with the PR #132 and enhance the caching with this PR later?
> 
> @shtripat @aruniiird
No issues... Will make needed changes to this, and we can put this in place on a later stage...